### DISCUSSION TITLE
Async methods which only call one async method

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#5722] [Q&A]
Imagine this:
```cs
Task Foo() { ... }
async Task Bar()
{
  // ...
  await Foo();
}
// or
Task Bar2()
{
  // ...
  return Foo();
}
```
Which one is preferable, `Bar` or `Bar2`? I imagine `Bar` has a performance overhead because it'll create the state machine, but the question is, is that considered a serious penalty to avoid? On the other hand, are there downsides to `Bar2`? Is debugging negatively impacted?

**Comment by CyrusNajmabadi** [comment]
Not a language question.  Note that this can very much affect things like how exceptions work.  Imagine if you have:

```c#
async Task Bar(string s!!)
{
  await Foo();
}
// or
Task Bar2(string s!!)
{
  return Foo();
}
```

In the first, if you pass null to `Bar` then you'll get a *Task* back in the faulted state.  In the second, when you call Bar2 you will synchronously get an ArgumentNullException back on the thread that called Bar2.

These are very different semantics and your code needs be careful here around how exception flow is managed.
